intel-npu-driver: init at 1.26.0

pseudocc opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Packaging the [Intel NPU](https://github.com/intel/linux-npu-driver) firmware (for Linux kernel module: intel_vpu).

```
[nix-shell:~/.cache/nixpkgs-review/pr-382756-4]$ tree results/intel-npu-driver-x86_64-linux/
results/intel-npu-driver-x86_64-linux/
├── bin
│   ├── npu-kmd-test
│   └── npu-umd-test
└── lib
    ├── libze_intel_vpu.so -> libze_intel_vpu.so.1
    ├── libze_intel_vpu.so.1 -> libze_intel_vpu.so.1.13.0
    └── libze_intel_vpu.so.1.13.0

3 directories, 5 files

[nix-shell:~/.cache/nixpkgs-review/pr-382756-4]$ tree results/intel-npu-firmware-x86_64-linux
results/intel-npu-firmware-x86_64-linux
└── lib
    └── firmware
        └── intel
            └── vpu
                ├── mtl_vpu_v0.0.bin -> vpu_37xx_v0.0.bin
                ├── vpu_37xx_v0.0.bin
                └── vpu_40xx_v0.0.bin

5 directories, 3 files
```

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

pseudocc commented:
I created a flake for it, and verified the firmware worked well.
https://github.com/intel/linux-npu-driver/pull/67
https://github.com/pseudocc/nix-config/commit/27ee1af101d08326150af267cdcfabce7de4365a

osbm commented on pkgs/by-name/in/intel-npu-firmware/package.nix:
```diff
@@ -0,0 +1,37 @@
+    owner = "intel";
+    repo = "linux-npu-driver";
+    rev = "bd414f6905d587bf4dd49a8f74511b7a5fd6f226";
+    sha256 = "ChiVnZ7SiiW5UIjYmvj9cf1bgS+GPHX279Q9DIA11og=";
```
also please use `hash = "sha256-[HASH]"` format

osbm commented on pkgs/by-name/in/intel-npu-firmware/package.nix:
```diff
@@ -0,0 +1,37 @@
+    runHook postInstall
+  '';
+
+  meta = with lib; {
```
Also we are trying to reduce the usage of `with lib;`

see https://github.com/NixOS/nixpkgs/issues/208242

osbm reviewed (changes requested):
I only have trivial change requests, package seems sufficiently up to standard otherwise.

jwludzik commented:
The NPU firmware is strictly connected with NPU compiler and driver. This is a problem, because whenever you want to use newer driver, ex. upcoming v1.14.0 release, then you should use the same firmware from the release as well. If the firmware is not matched with compiler, then we do not guarantee that all works correctly. Best would be to have a common package - firmware, driver and compiler. That is why the NPU firmware never reached the linux-firmware repository

SuperSandro2000 commented on nixos/modules/hardware/cpu/intel-npu.nix:
```diff
@@ -0,0 +1,26 @@
+  ...
+}:
+{
+  ###### interface
```
```suggestion
```

SuperSandro2000 commented on nixos/modules/hardware/cpu/intel-npu.nix:
```diff
@@ -0,0 +1,26 @@
+
+  };
+
+  ###### implementation
```
```suggestion
```

SuperSandro2000 commented on nixos/modules/hardware/cpu/intel-npu.nix:
I question this entire module. It just does to little to be added.

pseudocc replied:
I was just trying to get rid of the firmware loading error at the first place, while according to @jwludzik, other packages are also required to get NPU really work, I think it's good to keep the module to install all related packages together.

SuperSandro2000 commented on nixos/modules/installer/tools/nixos-generate-config.pl:
```diff
@@ -196,6 +196,11 @@ sub pciCheck {
         ($device eq "0x4229" || $device eq "0x4230" ||
          $device eq "0x4222" || $device eq "0x4227");
 
+    push @attrs, "hardware.cpu.intel.enableNpuFirmware = true;" if
```
I would recommend 

```suggestion
    push @attrs, "hardware.firmware = [ pkgs.intel-npu-firmware ];" if
```

pseudocc commented:
Thank you guys for reviewing this, I am changing this to WIP for packaging related driver and compiler as well.

pseudocc committed:
intel-npu-driver: init at 1.19.0

Only support the standalone build.

Closes: #348739

SuperSandro2000 commented on pkgs/by-name/in/intel-npu-firmware/package.nix:
```diff
@@ -0,0 +1,26 @@
+  inherit (intel-npu-driver) version src;
+  pname = "intel-npu-firmware";
+
+  installPhase = ''
```
Why do we need to compile everything again? why not just copy it to a separate output?

pseudocc replied:
I thought inheriting version and src will not cause building `intel-npu-driver`, let me re-organize this.

pseudocc replied:
Why I am not copying this to a separate output is that, [user may just install the firmware](https://github.com/NixOS/nixpkgs/pull/382756#discussion_r1978326188), and this approach will not need the cmake env since the firmware binary is in the source tree.

SuperSandro2000 replied:
Can we then set dontBuild to make that obvious?

I am not sure if that optimisation is worth it. Having two packages to update is some overhead and r-ryantm cannot deal with that properly, also when doing inherits.

Why not comine them again and move the firmware to the respecting output?

SuperSandro2000 commented on pkgs/by-name/in/intel-npu-driver/package.nix:
```diff
@@ -0,0 +1,79 @@
+    maintainers = with lib.maintainers; [ pseudocc ];
+  };
+
+  passthru = {
```
meta should always be last and be after passthru

SuperSandro2000 commented on pkgs/by-name/in/intel-npu-driver/package.nix:
```diff
@@ -0,0 +1,79 @@
+  passthru = {
+    inherit standalone;
+
+    level-zero = stdenv.mkDerivation {
```
What is the justification that we cannot use our system level-zero?

pseudocc replied:
somehow missed it, will remove this whole part.

SuperSandro2000 commented on pkgs/by-name/in/intel-npu-driver/package.nix:
```diff
@@ -0,0 +1,79 @@
+    level-zero = stdenv.mkDerivation {
+      inherit (baseDrv) src buildInputs nativeBuildInputs;
+      pname = "intel-npu-level-zero";
+      version = "${version}-1.18.5";
```
This will probably be forgotten to be updated...

SuperSandro2000 commented on pkgs/by-name/in/intel-npu-driver/package.nix:
```diff
@@ -0,0 +1,79 @@
+      installPhase = mkInstallPhase [ "level-zero" ];
+
+      meta = {
+        homepage = "https://github.com/oneapi-src/level-zero";
```
That is maybe a bit missleading

SuperSandro2000 commented on pkgs/by-name/in/intel-npu-driver/package.nix:
```diff
@@ -0,0 +1,79 @@
+    ];
+  };
+
+  mkInstallPhase = lib.concatMapStringsSep "\n" (
+    component: "cmake --install . --component ${component} --prefix $out"
+  );
```
We cannot use installTarget?

pseudocc replied:
I failed to find the make targets to do identical stuffs than this, I guess we just use this `--component` option here.

SuperSandro2000 commented on pkgs/by-name/in/intel-npu-driver/package.nix:
```diff
@@ -0,0 +1,79 @@
+  cmake,
+  git,
+  fetchFromGitHub,
+  standalone ? true,
```
What is that boolean even doing?

SuperSandro2000 commented on nixos/modules/hardware/cpu/intel-npu.nix:
```diff
@@ -0,0 +1,50 @@
+}:
+let
+  cfg = config.hardware.cpu.intel.npu;
+  type.driver =
+    with lib.types;
+    nullOr (enum [
+      "standalone"
+      # TODO: "with-compiler"
+    ]);
```
We should not add dead code for a feature that doesn't exist yet

SuperSandro2000 commented on nixos/modules/hardware/cpu/intel-npu.nix:
```diff
@@ -0,0 +1,50 @@
+  options = {
+    hardware.cpu.intel.npu = {
+      enable = lib.mkEnableOption "Intel NPU support";
+      driver = lib.mkOption {
+        type = type.driver;
+        default = "standalone";
+        description = ''
+          The driver to use for the Intel NPU.
+            - null: Only load the firmware.
+            - standalone: Use the standalone driver.
+
+          For more information, see:
+          https://github.com/intel/linux-npu-driver/blob/v1.13.0/docs/overview.md
+        '';
+      };
```
IMO if people only want to load the firmware, they can just add ``hardware.firmware = [ pkgs.intel-npu-firmware ];`` to their config. Lets remove this option for now.

SuperSandro2000 commented on nixos/modules/hardware/cpu/intel-npu.nix:
```diff
@@ -0,0 +1,50 @@
+          pkgs.intel-npu-driver
+          pkgs.intel-npu-driver.level-zero
+        ];
+        environment.sessionVariables.LD_LIBRARY_PATH = [
+          "${pkgs.intel-npu-driver}/lib"
+          "${pkgs.intel-npu-driver.level-zero}/lib"
+        ];
```
Is this absolutely necessary? LD_LIBRARY_PATH is pretty dirty and can cause all kinds of errors.

pseudocc replied:
the `npu-umd-test` will find `level-zero` in `LD_LIBRARY_PATH`, let me check if our system level-zero could be loaded, I see `addDriverRunpath $out/lib/libze_loader.so` in the fix-up phase.

pseudocc replied:
As long as I put `intel-npu-drive` in  `hardware.graphics.extraPackages`, `LD_LIBRARY_PATH` is not required to be set explicitly.

nmouha commented:
The driver proposed in #348739 seems to solve some of the issues that were pointed out.

It has just one file for both the driver and the firmware: `pkgs/by-name/in/intel-npu-driver/package.nix`

```
{ pkgs, ... }:
pkgs.stdenv.mkDerivation rec {
  pname = "intel-npu-driver";
  version = "1.13.0";

  src = pkgs.fetchFromGitHub {
    fetchSubmodules = true;
    fetchLFS = true;
    owner = "intel";
    repo = "linux-npu-driver";
    rev = "v${version}";
    hash = "sha256-+WPJrxwUT0UwU8VpJ4Wnmu/hLkdCDwiidGQwjl1Nvxk=";
  };
  outputs = [ "out" "firmware" ];
  postPatch = ''
    substituteInPlace firmware/CMakeLists.txt \
      --replace-fail "DESTINATION /lib/firmware/updates/intel/vpu/" \
      "DESTINATION $firmware/lib/firmware/intel/vpu/"
    substituteInPlace third_party/level-zero-npu-extensions/ze_graph_ext.h \
      --replace-fail "#include \"ze_api.h\"" "#include <level_zero/ze_api.h>"
  '';
  nativeBuildInputs = [ pkgs.cmake ];
  buildInputs = with pkgs; [
    udev
    boost
    openssl
    level-zero
  ];
  meta = {
    description = "Intel NPU (Neural Processing Unit) Driver";
    homepage = "https://github.com/intel/linux-npu-driver";
    license = pkgs.lib.licenses.mit;
  };
}
```

Which can be enabled from the NixOS configuration as follows:
```
  boot.extraModulePackages = [
    intel-npu-driver
  ];
  hardware.firmware = [
    intel-npu-driver.firmware
  ];
```

Hope this information can be helpful!

pseudocc commented on nixos/modules/hardware/cpu/intel-npu.nix:
```diff
@@ -0,0 +1,32 @@
+    ];
+
+    hardware.firmware = [ pkgs.intel-npu-firmware ];
+    hardware.graphics.extraPackages = [ pkgs.intel-npu-driver ];
```
@nmouha with this, user wouldn't need to pass `LD_LIBRARY_PATH` to the `npu-umd-test`.

nmouha replied:
Yes, nobody should be using `LD_LIBRARY_PATH`. I just wanted to point out that the some tests were failing in #348739, but they are now fixed.

Moreover, I'm not suggesting to provide `npu-kmd-test` and `umd-kmd-test` as commands by intel-npu-driver...

I was expecting that these would be [package tests](https://nixos.org/manual/nixpkgs/stable/#ssec-check-phase) (if compiled on a platform where the Intel NPU is available).

nilp0inter commented:
Hi @pseudocc,

I wanted to share an experiment I conducted to streamline the package compilation process as suggested by @SuperSandro2000. Here's what I did:

1. I added `"firmware"` to the list of outputs in the package configuration.

2. I modified the `postPatch` section to include the following:
   ```bash
   substituteInPlace firmware/CMakeLists.txt \                      
     --replace-fail "DESTINATION /lib/firmware/updates/intel/vpu/" \
     "DESTINATION $firmware/lib/firmware/intel/vpu/"
   ```

3. I removed the `installPhase`, compiling the whole package in a single step.

I hope this approach might be helpful as we work towards refining the packaging process.

pseudocc commented:
Hi @nilp0inter , I think that there might be a use case that the user just want the firmware only. If so, my implementation will not take any time preparing cmake env and building the artifacts since the firmware binary is in the source tree.

nilp0inter commented:
Thanks for the clarification @pseudocc . I was missing the point completely.

Cheers!

izelnakri commented:
Latest version of `npu-driver` is now v1.17 @pseudocc , lets update it and I don't want to hold back the PR from being merged. In fact, this PR has been on hold for 3 months for no good reason. Would appreciate if it gets approved & merged  right after the update @SuperSandro2000 @pseudocc

SuperSandro2000 commented on pkgs/by-name/in/intel-npu-firmware/package.nix:
```diff
@@ -0,0 +1,31 @@
+    rev = "v${version}";
+    hash = "sha256-ChiVnZ7SiiW5UIjYmvj9cf1bgS+GPHX279Q9DIA11og=";
+  };
+
```
```suggestion

  dontBuild = true;

```

SuperSandro2000 commented on pkgs/by-name/in/intel-npu-driver/package.nix:
```diff
@@ -0,0 +1,63 @@
+    owner = "intel";
+    repo = "linux-npu-driver";
+    rev = "v${version}";
+    fetchSubmodules = true;
```
Vendoring libraries we already have in nixpkgs is not nice.

SuperSandro2000 commented on pkgs/by-name/in/intel-npu-driver/package.nix:
```diff
@@ -0,0 +1,63 @@
+  ];
+
+  postPatch = ''
+    for test in kmd-test umd-test; do
+      substituteInPlace validation/$test/CMakeLists.txt --replace \
+        "COMPONENT validation-npu" \
+        "DESTINATION $validation/bin COMPONENT validation-npu"
+    done
```
```suggestion    
    substituteInPlace validation/{kmd-test,umd-test}/CMakeLists.txt --replace \
      "COMPONENT validation-npu" \
      "DESTINATION $validation/bin COMPONENT validation-npu"
```

pseudocc replied:
Applied.

pseudocc commented:
Ran `npu-umd-test` and `npu-kmd-test` looked fine on my ThinkPad X13 Yoga Gen5 (CPU: Intel(R) Core(TM) Ultra 5 135U).

```
$ patchelf --print-rpath $(which npu-umd-test
/nix/store/g4h8yviv4r2rfnwzlmc746s5ls22zvgd-level-zero-1.22.4/lib:/nix/store/q4wq65gl3r8fy746v9bbwgx4gzn0r2kl-glibc-2.40-66/lib:/nix/store/l7d6vwajpfvgsd3j4cr25imd1mzb7d1d-gcc-14.3.0-lib/lib
```

pseudocc commented on pkgs/by-name/in/intel-npu-driver/package.nix:
```diff
@@ -0,0 +1,63 @@
+    udev
+    openssl
+    boost
+    level-zero
```
I removed it in the postPatch section
```
rm third_party/cmake/level-zero.cmake
``

izelnakri commented:
@SuperSandro2000 @osbm seems like this PR is more than 6 months old already and requested changes are made. Could you approve your reviews and lets merge this? 

I've been custom configuring this to my config to utilize my NPU for far too long. Thanks!

NeuerUser commented:
> I've been custom configuring this to my config to utilize my NPU for far too long. Thanks!

Could you tell me how you did the custom config? I've been waiting for this PR for about six months now and would really like to use the NPU.

Thanks 🙏

pseudocc commented:
> > I've been custom configuring this to my config to utilize my NPU for far too long. Thanks!
> 
> Could you tell me how you did the custom config? I've been waiting for this PR for about six months now and would really like to use the NPU.
> 
> Thanks 🙏

@NeuerUser You may refer to mine (if you are also using flake): https://github.com/pseudocc/nix-config/commit/6434c09d545663430f11e49775c76193e9163b3c

pseudocc committed:
intel-npu-driver: update to 1.22.0

Part-Of: #348739

NeuerUser commented:
> @NeuerUser You may refer to mine (if you are also using flake): [pseudocc/nix-config@6434c09](https://github.com/pseudocc/nix-config/commit/6434c09d545663430f11e49775c76193e9163b3c)

Thanks a lot. I do, however, get an SHA error:

`

         specified: sha256-GxmAmbf28hn/ragkTkFBYbM6Z+uFR6X4foMOyE8UVAc=
            got:    sha256-e1pkwFXfnf9NyjAUGrJJB5r1ytI1l6X0X5GF63O868U=
`

dmivankov commented:
Just as a datapoint, on 25.05 I managed to run validation without applying the module

```console
$ LD_LIBRARY_PATH="$(nix-build -A intel-npu-driver.out)/lib" "$(nix-build -A intel-npu-driver.validation)/bin/npu-umd-test"

...
[ RUN      ] Umd.ConfigurationCheck
/build/source/validation/umd-test/umd_test.cpp:201: Failure
Expected: (config.size()) > (0), actual: 0 vs 0
Missed configuration file.
Test scope is reduced.
Provide valid configurtion file, use:
npu-umd-test -c/--config [CONFIGURATION_PATH]

[  FAILED  ] Umd.ConfigurationCheck (0 ms)
...
[ DISABLED ] MemoryExecution.DISABLED_CopyingFromUnpinnedHostMemoryShouldBeAllowed
...
```
all other tests are either passing or skipped

And for kmd
```console
$ LD_LIBRARY_PATH="$(nix-build -A intel-npu-driver.out)/lib" sudo "$(nix-build -A intel-npu-driver.validation)/bin/npu-kmd-test"

...
[ RUN      ] Device.ResetComputeEngine
/build/source/validation/kmd-test/test_device.cpp:118: Failure
Expected equality of these values:
  write_debugfs_file("reset_engine", 0)
    Which is: 19
  0

[  FAILED  ] Device.ResetComputeEngine (3 ms)
[ DISABLED ] Device.DISABLED_ResetInvalidEngine
...
[ DISABLED ] UapiMem.DISABLED_BoClosedDuringCreate
...
```
and all other passing

pseudocc committed:
intel-npu-driver: update to 1.23.0

Part-Of: #348739

SuperSandro2000 reviewed (commented):
LGTM with those two notes

pseudocc committed:
intel-npu-driver: apply suggestions

pseudocc commented on nixos/modules/hardware/cpu/intel-npu.nix:
```diff
@@ -0,0 +1,32 @@
+      pkgs.intel-npu-driver.validation
+      pkgs.level-zero
+    ];
+  };
```
applied.

pseudocc commented on pkgs/by-name/in/intel-npu-driver/package.nix:
```diff
@@ -0,0 +1,75 @@
+  src = fetchFromGitHub {
+    owner = "intel";
+    repo = "linux-npu-driver";
+    rev = "v${version}";
```
applied.

pseudocc committed:
intel-npu-driver: update to 1.24.0

osbm reviewed (approved):
LGTM!!

Thank you for your contributions. <3

pseudocc committed:
intel-npu-driver: update to 1.26.0

Fix building the third party library yaml-cpp:
- Compatibility with CMake < 3.5 has been removed from CMake.
